### PR TITLE
Introduce generics to splay

### LINKED ISSUE #304: Update Go version to 1.18
**Description**:

Update Go version to 1.18.

We need to check about changes from 1.17 to 1.18 and reflect the changes on the Yorkie.

https://tip.golang.org/doc/go1.18

Tasks

- [x] https://github.com/yorkie-team/yorkie/issues/234
- [x] Introduce Generics
  - [x] [cache.go](https://github.com/yorkie-team/yorkie/blob/main/pkg/cache/cache.go)
  - [x] [splay.go](https://github.com/yorkie-team/yorkie/blob/main/pkg/splay/splay.go)
  - [x] [priority_queue](https://github.com/yorkie-team/yorkie/blob/main/pkg/pq/priority_queue.go)
  - [x] [llrb.go](https://github.com/yorkie-team/yorkie/blob/main/pkg/llrb/llrb.go)
  - [x] [rga_tree_split](https://github.com/yorkie-team/yorkie/blob/main/pkg/document/json/rga_tree_split.go)

**Why**:

- We can reflect on the latest Go improvements and keep an up-to-date code.

### PR DESCRIPTION
**Pull request by sejongk**
**What this PR does / why we need it**:

Introduce generics to splay

**Which issue(s) this PR fixes**:

Fixes #304 

**Special notes for your reviewer**:

By introducing generics to splay, we don't have to use the type assertions in RGATreeSplit and RGATreeList. 
If there is anything that needs improvement, please let me know. :)

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation**:

```docs

```

**Checklist**:
- [X] Added relevant tests or not required
- [X] Didn't break anything

### CODE REVIEW

**Comment by hackerwins** [comment]
@sejongk Thanks for your contribution.
Please check the lint errors on the CI.

File: pkg/splay/splay.go [hunk 1]
```diff
@@ -28,77 +28,77 @@ type Value interface {
...
+	n := &Node[V]{
 		value: value,
 	}
 	n.initWeight()
 	return n
 }
 
 // Value returns the value of this Node.
-func (n *Node) Value() Value {
+func (n *Node[V]) Value() V {
```
Comment (hackerwins) [inline pkg/splay/splay.go:50]:
```suggestion
func (t *Node[V]) Value() V {
```

File: pkg/splay/splay.go [hunk 2]
```diff
@@ -28,77 +28,77 @@ type Value interface {
...
 	return n.value
 }
 
-func (n *Node) leftWeight() int {
+func (n *Node[V]) leftWeight() int {
```
Comment (hackerwins) [inline pkg/splay/splay.go:54]:
```suggestion
func (t *Node[V]) leftWeight() int {
```

File: pkg/splay/splay.go [hunk 3]
```diff
@@ -28,77 +28,77 @@ type Value interface {
...
 	if n.left == nil {
 		return 0
 	}
 	return n.left.weight
 }
 
-func (n *Node) rightWeight() int {
+func (n *Node[V]) rightWeight() int {
```
Comment (hackerwins) [inline pkg/splay/splay.go:61]:
```suggestion
func (t *Node[V]) rightWeight() int {
```

File: pkg/splay/splay.go [hunk 4]
```diff
@@ -28,77 +28,77 @@ type Value interface {
...
 	if n.right == nil {
 		return 0
 	}
 	return n.right.weight
 }
 
-func (n *Node) initWeight() {
+func (n *Node[V]) initWeight() {
```
Comment (hackerwins) [inline pkg/splay/splay.go:68]:
```suggestion
func (t *Node[V]) initWeight() {
```

File: pkg/splay/splay.go [hunk 5]
```diff
@@ -28,77 +28,77 @@ type Value interface {
...
 	n.weight = n.value.Len()
 }
 
-func (n *Node) increaseWeight(weight int) {
+func (n *Node[V]) increaseWeight(weight int) {
```
Comment (hackerwins) [inline pkg/splay/splay.go:72]:
```suggestion
func (t *Node[V]) increaseWeight(weight int) {
```

File: pkg/splay/splay.go [hunk 6]
```diff
@@ -28,77 +28,77 @@ type Value interface {
...
 	n.weight += weight
 }
 
-func (n *Node) unlink() {
+func (n *Node[V]) unlink() {
```
Comment (hackerwins) [inline pkg/splay/splay.go:76]:
```suggestion
func (t *Node[V]) unlink() {
```

File: pkg/splay/splay.go [hunk 7]
```diff
@@ -28,77 +28,77 @@ type Value interface {
...
 	n.parent = nil
 	n.right = nil
 	n.left = nil
 }
 
-func (n *Node) hasLinks() bool {
+func (n *Node[V]) hasLinks() bool {
```
Comment (hackerwins) [inline pkg/splay/splay.go:82]:
```suggestion
func (t *Node[V]) hasLinks() bool {
```

**Review by hackerwins — COMMENTED** [review]
Looking at the lint errors, it is thought to be the lint malfunction. 🤔
I changed the Node to the same receiver name as Tree to avoid it for now.

File: pkg/splay/splay.go [hunk 8]
```diff
@@ -28,77 +28,77 @@ type Value interface {
...
 		value: value,
 	}
 	n.initWeight()
 	return n
 }
 
 // Value returns the value of this Node.
-func (n *Node) Value() Value {
+func (t *Node[V]) Value() V {
 	return n.value
```
Comment (hackerwins) [inline pkg/splay/splay.go:51]:
```suggestion
	return t.value
```

File: pkg/splay/splay.go [hunk 9]
```diff
@@ -28,77 +28,77 @@ type Value interface {
...
 }
 
-func (n *Node) leftWeight() int {
+func (t *Node[V]) leftWeight() int {
 	if n.left == nil {
```
Comment (hackerwins) [inline pkg/splay/splay.go:55]:
```suggestion
	if t.left == nil {
```

File: pkg/splay/splay.go [hunk 10]
```diff
@@ -28,77 +28,77 @@ type Value interface {
...
 		return 0
 	}
 	return n.left.weight
```
Comment (hackerwins) [inline pkg/splay/splay.go:58]:
```suggestion
	return t.left.weight
```

File: pkg/splay/splay.go [hunk 11]
```diff
@@ -28,77 +28,77 @@ type Value interface {
...
 }
 
-func (n *Node) rightWeight() int {
+func (t *Node[V]) rightWeight() int {
 	if n.right == nil {
```
Comment (hackerwins) [inline pkg/splay/splay.go:62]:
```suggestion
	if t.right == nil {
```

File: pkg/splay/splay.go [hunk 12]
```diff
@@ -28,77 +28,77 @@ type Value interface {
...
 		return 0
 	}
 	return n.right.weight
```
Comment (hackerwins) [inline pkg/splay/splay.go:65]:
```suggestion
	return t.right.weight
```

File: pkg/splay/splay.go [hunk 13]
```diff
@@ -28,77 +28,77 @@ type Value interface {
...
 }
 
-func (n *Node) initWeight() {
+func (t *Node[V]) initWeight() {
 	n.weight = n.value.Len()
```
Comment (hackerwins) [inline pkg/splay/splay.go:69]:
```suggestion
	n.weight = t.value.Len()
```

File: pkg/splay/splay.go [hunk 14]
```diff
@@ -28,77 +28,77 @@ type Value interface {
...
 }
 
-func (n *Node) increaseWeight(weight int) {
+func (t *Node[V]) increaseWeight(weight int) {
 	n.weight += weight
```
Comment (hackerwins) [inline pkg/splay/splay.go:73]:
```suggestion
	t.weight += weight
```

File: pkg/splay/splay.go [hunk 15]
```diff
@@ -28,77 +28,77 @@ type Value interface {
...
 }
 
-func (n *Node) unlink() {
+func (t *Node[V]) unlink() {
 	n.parent = nil
 	n.right = nil
 	n.left = nil
```
Comment (hackerwins) [inline pkg/splay/splay.go:79]:
```suggestion
	t.parent = nil
	t.right = nil
	t.left = nil
```

File: pkg/splay/splay.go [hunk 16]
```diff
@@ -28,77 +28,77 @@ type Value interface {
...
 }
 
-func (n *Node) hasLinks() bool {
+func (t *Node[V]) hasLinks() bool {
 	return n.parent != nil || n.left != nil || n.right != nil
```
Comment (hackerwins) [inline pkg/splay/splay.go:83]:
```suggestion
	return t.parent != nil || t.left != nil || t.right != nil
```

File: pkg/splay/splay.go [hunk 17]
```diff
@@ -28,77 +28,77 @@ type Value interface {
...
 		return 0
 	}
-	return n.right.weight
+	return t.right.weight
 }
 
-func (n *Node) initWeight() {
-	n.weight = n.value.Len()
+func (t *Node[V]) initWeight() {
+	n.weight = t.value.Len()
```
Comment (hackerwins) [inline pkg/splay/splay.go:69]:
```suggestion
	t.weight = t.value.Len()
```

**Review by hackerwins — APPROVED** [review]

**Comment by sejongk** [comment]
@hackerwins Thanks for your edit.

> Looking at the lint errors, it is thought to be the lint malfunction.
I'm totally agree with you. I will check whether the linter is working properly and share related information if the current linter's upgrade is necessary.

**Comment by sejongk** [comment]
@hackerwins It seems a bug in the `revive` module enabled in golangci-lint, and the bug is fixed (ref. https://github.com/mgechev/revive/pull/692). 
When golangci-lint bumps up its `revive` to the new version that includes the patch, we may resolve the above problem by upgrading golangci-lint.

**Pull request merged**